src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/TempPE/DesignTimeInputsChangeTracker.cs

```diff
@@ -0,0 +1,238 @@
+        /// This is to allow unit tests to force completion of our source block rather than waiting for async work to complete
+        /// </summary>
+        internal bool AllowSourceBlockCompletion { get; set; }
+
+        public override NamedIdentity DataSourceKey { get; } = new NamedIdentity(nameof(DesignTimeInputsChangeTracker));
+
+        public override IComparable DataSourceVersion => _version;
+
+        public override IReceivableSourceBlock<IProjectVersionedValue<DesignTimeInputsDelta>> SourceBlock
+        {
+            get
+            {
+                EnsureInitialized();
+
+                return _publicBlock!;
+            }
+        }
+
+        protected override void Initialize()
+        {
+            base.Initialize();
+
+            // Create an action block to process the design time inputs and configuration general changes
+            _inputsActionBlock = DataflowBlockSlim.CreateActionBlock<IProjectVersionedValue<Tuple<DesignTimeInputs, IProjectSubscriptionUpdate>>>(ProcessDataflowChanges);
+
+            _broadcastBlock = DataflowBlockSlim.CreateBroadcastBlock<IProjectVersionedValue<DesignTimeInputsDelta>>(nameFormat: nameof(DesignTimeInputsChangeTracker) + "Broadcast {1}");
+            _publicBlock = AllowSourceBlockCompletion ? _broadcastBlock : _broadcastBlock.SafePublicize();
+
+            IDisposable projectLink = ProjectDataSources.SyncLinkTo(
+                   _inputsDataSource.SourceBlock.SyncLinkOptions(
+                       linkOptions: DataflowOption.PropagateCompletion),
+                   _projectSubscriptionService.ProjectRuleSource.SourceBlock.SyncLinkOptions(
+                      linkOptions: DataflowOption.WithRuleNames(ConfigurationGeneral.SchemaName)),
+                   _inputsActionBlock,
+                   DataflowOption.PropagateCompletion,
+                   cancellationToken: _project.Services.ProjectAsynchronousTasks.UnloadCancellationToken);
+
+            // Create an action block to process file change notifications
+            _fileWatcherActionBlock = DataflowBlockSlim.CreateActionBlock<IProjectVersionedValue<string[]>>(ProcessFileChangeNotification);
+            IDisposable watcherLink = _fileWatcher.SourceBlock.LinkTo(_fileWatcherActionBlock, DataflowOption.PropagateCompletion);
+
+            _disposables.AddDisposable(projectLink);
+            _disposables.AddDisposable(watcherLink);
+
+            JoinUpstreamDataSources(_inputsDataSource, _projectSubscriptionService.ProjectRuleSource, _fileWatcher);
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            if (_inputsActionBlock != null)
+            {
+                // This will stop our blocks taking any more input
+                _inputsActionBlock.Complete();
+                _fileWatcherActionBlock!.Complete();
+
+                _threadingService.ExecuteSynchronously(() => Task.WhenAll(_inputsActionBlock.Completion, _fileWatcherActionBlock.Completion));
```

Comment:
Why do we need this? This says "block a threadpool thread while we wait for these blocks to finish" - but you aren't using the results.

Reply:
We need to wait for completion to ensure that the disposal doesn't overlap with processing more input, as Complete is not a blocking call. This was the design we worked out for the file watcher, copied from somewhere in CPS (I don't remember where): https://github.com/dotnet/project-system/blob/feature/TempPE/src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/TempPE/DesignTimeInputsFileWatcher.cs#L160

If this case is different somehow I'd appreciate any extra context you can share.

Reply:
Dispose doesn't do anything other than complete the blocks, what's the race? File watcher dispose uses the results that the data flow listener touches, so the case is different.

Reply:
I'll whiteboard for you on Monday how this occurs.

Reply:
Fair enough, I took it to be a general problem to solve. Thanks for the explanation.

> I'll whiteboard for you on Monday how this occurs.

That would be awesome, there are still some comments from the previous review that I wanted to talk through with you a few weeks ago.